### PR TITLE
#6914 Multiplayer Status Display

### PR DESCRIPTION
**Pull request by Azzurite**
Implements #6914
Implements #5756

Alrighty, time for the fun stuff!

Look at the issue description for a description of the feature this PR adds. Comments about the code follow here.

To review this, I'd heartily recommend to do it commit-by-commit. I've neatly extracted almost each little step into their own commit, so the final commit that adds this feature is actually relatively small. I'll also be grouping my comments by commit.

* 50e22d9a0ff3950c3984e288ce495be2c7d9ce17 Handle subclassing of Events properly.
    This adds support for subclassing of events to the `EventBus`. What this means is that you'll be able to do this:
   ```
   open class Parent : Event
   class Child : Parent
   events.receive(Parent::class) { println(42) }
   EventBus.send(Child()) // will print 42
   ```
* b3d2b50a093c98f556f4c27266b25c69332512f1 Add relevant parent classes for the multiplayer events, this already enables a small code duplication removal in the `MultiplayerScreen`
* fccb7ec6eaa7d14089038a43d1486bc161fc7d6f Add being able to stop listening to events in the EventBus
   This can eventually be changed to be able to remove specific listeners, but that wasn't necessary in this PR yet.
* 1cd3072f92586bd3714164866321d295a493633c Add tests for EventBus: I had a bug at the beginning where, when I added the subclassing support, I had the same event arriving multiple times.
* f0cd85522c96b33b9ad665d4c29c2ba8e9630dab Make loading a multiplayer game automatically update the in-memory game in OnlineMultiplayer
   This is for saved games of multiplayer games, but also in general. Whenever you load a multiplayer game from somewhere, this updates the `savedGames` map in `OnlineMultiplayer` with a current preview, just to save some loads.
* bcff1b5b5aa7e5d1ddeb1a81fa338266018a89fb Refactor: Extract multiplayer settings into separate object
   I wanted to migrate the `multiplayerTurnCheckerDelayInMinutes` from a `Long` to a `java.time.Duration` anyway, and I decided to split the multiplayer settings off from the main settings object while I'm at it. Also includes a very rudimentary settings migration system: just check which `version` is defined in the settings file and do migrations depending on that. Currently only one migration of course from `null` version to `1`. When we then make further changes in the settings, we can just increment `version` and add another function in `GameSettingsMigrations`.
* 0ccafaf8c744036cc5dee27f87c6c64c35383296 Add multiplayer status display
   The meat of the whole thing here. This adds the features described in the linked issue.
* 5b8b6c30884dbbf2abe51ffc96834f7cfceb5f37  Fix error with multiplayer games not correctly being cleaned up after successful update
   In the multiplayer game list, when an error happened (game preview could not be downloaded), the "error icon" did not get cleaned up correctly. When what was returned remotely was the same as was contained in the file, an `MultiplayerGameUpdateUnchanged` got sent, but that wasn't handled in the `GameLIst`. So the game had a "error icon" even though it didn't have an error anymore.
* cb65c9b47d6c1500c7b359286b2fde2e0c924f60 Prevent loadLatestMultiplayerState() while next turn update is running
   Until now, whenever you pressed "Next turn", the game immediately downloaded the game you just uploaded, to update the current game state. But since you just uploaded this game state, there's no need to download it immediately again.
* 6c4ee580b97e4f3e04037447d9ad95c1c15b71cd Show "Working..." while waiting for next turn calculations instead of "Waiting for [civ]..."
   Of course this is not related to this feature... but I saw no reason to extract these 2 lines from this PR into another one

**Known Issue**
* ~~In some very rare cases when you're playing multiple games, the current game does not update when another player finishes their turn and it goes back to your turn. Still investigating.~~ - fixed by bc7865c49595bf85a1e7c70de3ace18d55b674bd

### CODE REVIEW

File: android/assets/jsons/translations/template.properties [hunk 1]
```diff
@@ -1342,7 +1344,7 @@ Choose name for [unitName] =
 # Multiplayer Turn Checker Service
 
 Enable out-of-game turn notifications = 
-Time between turn checks out-of-game (in minutes) = 
+Out-of-game, update status of all games every: =
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply android/assets/jsons/translations/template.properties:1349]: There's no right way, that's all. Or, that there _is_ the right way, and translators will lose the old translation and need to start from scratch - or retrieve from git history. There _was_ an attempt to alleviate that, no longer.

File: core/src/com/unciv/ui/options/MultiplayerTab.kt [hunk 2]
```diff
@@ -134,19 +170,86 @@ private fun fixTextFieldUrlOnType(TextField: TextField) {
...
+    val arr = Array<T>(size)
+    for (it in this) {
+        arr.add(it)
+    }
+    return arr
+}
+
+private fun usesDropbox(settings: GameSettings) = settings.multiplayer.server == Constants.dropboxMultiplayerServer
+
+private fun addRefreshSelect(
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/ui/options/MultiplayerTab.kt:208]: This leaves two cells in destination `table`, but calling code doesn't 'grok' there's now two columns where there was one, and other controls on the page lack the colspan... Noo biggie

**Review by SomeTroglodyte — APPROVED** [review]
No horrors from the depth on a test run. I'm no expert on MultiPlayer, but my :game_die: say :+1:

**Comment by yairm210** [comment]
Resolve conflicts?

**Comment by touhidurrr** [comment]
It seems like some atlas files were tweaked in resolve conflict stuff.
I have always wonder if these files should even be add to the repo.
As I saw them rebuilding themselves on device storage everytime something has changed.
if they are going to be built anyways then why add them? Possible waste of bandwidth.

**Comment by SomeTroglodyte** [comment]
And how then will the textures get to Android users?

**Comment by touhidurrr** [comment]
maybe they will build themselves on Andorid too? needs testing though.
I saw those rebuilding themselves when I launched Unciv in my Pi.

**Comment by touhidurrr** [comment]
This happens when I deleted game.atlas and game.png on **Brave New World** and **5Hex Tileset** Mod.
```bash
~$ java -jar Unciv.jar
Reading: /home/admin/Unciv/mods/Brave New World/Images
UnitPromotionIcons/Buffalo Horns (alias of UnitPromotionIcons/Buffalo Chest)
UnitPromotionIcons/Buffalo Loins (alias of UnitPromotionIcons/Buffalo Chest)
BuildingIcons/Cloves (alias of ResourceIcons/Cloves)
UnitPromotionIcons/Heroism (alias of UnitPromotionIcons/Ambition)
UnitPromotionIcons/Homeland Guardian (alias of UnitPromotionIcons/Desert Warrior)
UnitPromotionIcons/Invulnerability (alias of UnitPromotionIcons/Ambition)
UnitPromotionIcons/Mystic Blade (alias of UnitPromotionIcons/Ambition)
BuildingIcons/Nutmeg (alias of ResourceIcons/Nutmeg)
BuildingIcons/Pepper (alias of ResourceIcons/Pepper)
UnitPromotionIcons/Restlessness (alias of UnitPromotionIcons/Ambition)
UnitPromotionIcons/Sneak Attack (alias of UnitPromotionIcons/Ambition)
Packing.........
Writing 2048x512: mods/Brave New World/game.png
Reading: /home/admin/Unciv/mods/5Hex Tileset/Images
TileSets/5Hex/Tiles/AncientRuinsForests (alias of TileSets/5Hex/Tiles/Ancient ruins)
TileSets/5Hex/Tiles/City center-Classical era (alias of TileSets/5Hex/Tiles/City center-Ancient era)
TileSets/5Hex/Tiles/City center-Future era (alias of TileSets/5Hex/Tiles/City center-Atomic era)
TileSets/5Hex/Tiles/City center-Information era (alias of TileSets/5Hex/Tiles/City center-Atomic era)
TileSets/5Hex/Tiles/City center-Modern era (alias of TileSets/5Hex/Tiles/City center-Industrial era)
TileSets/5Hex/Tiles/City center-Renaissance era (alias of TileSets/5Hex/Tiles/City center-Medieval era)
TileSets/5Hex/Tiles/Land Trade Route (Gold) (alias of TileSets/5Hex/Tiles/Land Trade Route (Food))
TileSets/5Hex/Tiles/Land Trade Route (Production) (alias of TileSets/5Hex/Tiles/Land Trade Route (Food))
TileSets/5Hex/Tiles/Sea Trade Route (Gold) (alias of TileSets/5Hex/Tiles/Sea Trade Route (Food))
TileSets/5Hex/Tiles/Sea Trade Route (Production) (alias of TileSets/5Hex/Tiles/Sea Trade Route (Food))
Packing.........
Packing.........
Packing.........
Packing.........
Writing 2048x2048: mods/5Hex Tileset/game.png
Writing 2048x2048: mods/5Hex Tileset/game2.png
Writing 2048x2048: mods/5Hex Tileset/game3.png
Writing 2048x512: mods/5Hex Tileset/game4.png
Packing textures - 45583ms
Could not initialize Discord
[ALSOFT] (EE) Failed to set real-time priority for thread: Operation not permitted (1)
[ALSOFT] (EE) Failed to set real-time priority for thread: Operation not permitted (1)
Loading Construction = Construction.atlas
Loading Flags = Flags.atlas
Loading NationIcons = NationIcons.atlas
Loading PolicyIcons = PolicyIcons.atlas
Loading ReligionIcons = ReligionIcons.atlas
Loading Skin = Skin.atlas
Loading Tech = Tech.atlas
Loading UnitPromotionIcons = UnitPromotionIcons.atlas
Loading 5Hex Tileset = mods/5Hex Tileset/game.atlas
Loading ruleset - 2291ms
Loading ruleset - 3078ms
Loading ruleset - 789ms
Mod loaded successfully: Brave New World

Loading ruleset - 6ms
Mod loaded successfully: 5Hex Tileset

Loading percent complete of languages - 8ms
TileSetConfig loaded successfully: HexaRealm.json

TileSetConfig loaded successfully: FantasyHex.json

TileSetConfig loaded successfully: Default.json

TileSetConfig loaded successfully: mods/5Hex Tileset/jsons/TileSets/5Hex.json
```
And those files rebuilt themselves on first run.

**Comment by touhidurrr** [comment]
Well I change my mind. Cause it takes quite a while for them to be rebuilt. ~ < 1 min though.

**Pull request merged**